File: exampleBots/java/Perceptron.java
import java.util.ArrayList;
import java.util.Random;
import java.util.List;

public class Perceptron {

    private ArrayList<Double> weights;
    private static final Random rand = new Random();

    // constructor to initialize weights with random values between -1 and 1
    public Perceptron(int numInputs) {
        weights = new ArrayList<>();
        for (int i = 0; i <= numInputs; i++) {
            weights.add(rand.nextDouble() * 2 - 1); // random value between -1 and 1
        }
    }

    // set the weights of the perceptron manually
    public void setWeights(ArrayList<Double> newWeights) {
        weights.clear();
        weights.addAll(newWeights);
    }

    // predict the output for given inputs
    public int predict(ArrayList<Integer> inputs) {
        inputs.add(-1); // constant -1 input for the threshold weight (bias)
        
        // the dot product between the weights and the inputs
        double weightedSum = 0;
        for (int i = 0; i < weights.size(); i++) {
            weightedSum += weights.get(i) * inputs.get(i);
        }
        return activation(weightedSum);
    }

    // activation function to determine output
    private int activation(double x) {
        return x > 0 ? 1 : 0; // if weighted sum positive, output 1 otherwise output 0
    }

    // train the perceptron using the given inputs and desired outcomes
    public void train(ArrayList<ArrayList<Integer>> trainingInputs, ArrayList<Integer> outcomes, double learningRate, int rounds) {
        for (int round = 0; round < rounds; round++) {
            for (int i = 0; i < trainingInputs.size(); i++) {
                ArrayList<Integer> inputs = new ArrayList<>(trainingInputs.get(i));
                int expectedOutcome = outcomes.get(i);
                int prediction = predict(inputs);
                int error = expectedOutcome - prediction; // compute error
                inputs.add(-1); // add bias input for the weight update

                // adjust each weight in the direction to reduce the error using the perceptron rule
                for (int j = 0; j < weights.size(); j++) {
                    weights.set(j, weights.get(j) + learningRate * error * inputs.get(j));
                }
            }
        }
    }

    public ArrayList<Double> getWeights() {
        return new ArrayList<>(weights);
    }

    public static void main(String[] args) {
        // xreate all 32 possible inputs for a 5 input binary perceptron
        ArrayList<ArrayList<Integer>> trainingInputs = new ArrayList<>();
        for (int i = 0; i < 32; i++) {
            ArrayList<Integer> bits = new ArrayList<>();
            String bitString = String.format("%05d", Integer.parseInt(Integer.toBinaryString(i)));
            for (char c : bitString.toCharArray()) {
                bits.add(Character.getNumericValue(c));
            }
            trainingInputs.add(bits);
        }

        // generate desired outputs based on the specified input patterns
        ArrayList<Integer> desiredOutputs = new ArrayList<>();
        for (ArrayList<Integer> inp : trainingInputs) {
            if (inp.equals(new ArrayList<Integer>(List.of(0, 1, 1, 1, 1))) ||
                    inp.equals(new ArrayList<Integer>(List.of(1, 0, 1, 1, 1))) ||
                    inp.equals(new ArrayList<Integer>(List.of(1, 1, 0, 1, 1))) ||
                    inp.equals(new ArrayList<Integer>(List.of(1, 1, 1, 0, 1))) ||
                    inp.equals(new ArrayList<Integer>(List.of(1, 1, 1, 1, 0))) ||
                    inp.equals(new ArrayList<Integer>(List.of(1, 1, 1, 1, 1)))) {
                desiredOutputs.add(1);
            } else {
                desiredOutputs.add(0);
            }
        }

        ArrayList<Double> manualWeights = new ArrayList<>(List.of(0.1, 0.3, 0.4, 0.3, 0.3, 1.0));

        // test the perceptron with manually set weights.
        Perceptron perceptron = new Perceptron(5);
        perceptron.setWeights(manualWeights);
        System.out.println("Testing with manually set weights...");
        for (int i = 0; i < trainingInputs.size(); i++) {
            ArrayList<Integer> inputs = new ArrayList<>(trainingInputs.get(i));
            int desiredOutput = desiredOutputs.get(i);
            System.out.println("Input: " + inputs + " -> Predicted Output: " + perceptron.predict(inputs) + ", Desired Output: " + desiredOutput);
        }

        // train the perceptron and then test with the trained weights.
        Perceptron trainedPerceptron = new Perceptron(5);
        trainedPerceptron.train(trainingInputs, desiredOutputs, 0.1, 1000);
        System.out.println("\nTesting with trained weights...");
        for (int i = 0; i < trainingInputs.size(); i++) {
            ArrayList<Integer> inputs = new ArrayList<>(trainingInputs.get(i));
            int desiredOutput = desiredOutputs.get(i);
            System.out.println("Input: " + inputs + " -> Predicted Output: " + trainedPerceptron.predict(inputs) + ", Desired Output: " + desiredOutput);
        }

        // print the final weights after training.
        System.out.println("\nFinal weights: " + trainedPerceptron.getWeights());
    }
}

File: perceptron/perceptron.py
# Russell Kosovsky 9/22/23

import random

class Perceptron:
    
    def __init__(self, num_inputs):
        self.weights = []
        # append weights as random values between -1 and 1.
        # num_inputs + 1 because of the weight for bias (threshold).
        for _ in range(num_inputs + 1):
           self.weights.append(random.uniform(-1, 1))
        
    def set_weights(self, new_weights):
        # set the weights of the perceptron manually.
        self.weights = new_weights[:]

    def predict(self, inputs):
        # constant -1 input for the threshold weight.
        inputs.append(-1)
        # this is the dot product between the weights and the inputs.
        weighted_sum = 0
        for i in range(len(self.weights)):
            weighted_sum += self.weights[i] * inputs[i]
        # the activation function determines the output. 
        return self.activation(weighted_sum)

    def activation(self, x):
        # if the weighted sum is positive, output 1, otherwise, output 0.
        if x > 0:
            return 1
        else:
            return 0

    def train(self, training_inputs, outcome, learning_rate = 0.1, rounds = 1000):
        for round in range(rounds):
            for i in range(len(training_inputs)):
                # copy the input to avoid altering the original list.
                inputs = training_inputs[i][:]
                # extract the corresponding outcomes for the input.
                expected_outcome = outcome[i]
                # make a prediction using the current weights.
                prediction = self.predict(inputs)
                # compute the error: difference between label and prediction.
                error = expected_outcome - prediction
                # add bias input for the weight update.
                inputs.append(-1)
                # adjust each weight in the direction to reduce the error using the perceptron rule.
                for j in range(len(self.weights)):
                    self.weights[j] += learning_rate * error * inputs[j]


if __name__ == "__main__":
    
    # generate all 32 possible inputs for a 5-input binary perceptron.
    training_inputs = []
    for i in range(32):
        bits = [int(b) for b in format(i, '05b')]
        training_inputs.append(bits)

    # generate desired outputs based on the specified input patterns.
    desired_outputs = []
    for inp in training_inputs:
        if inp in [[0,1,1,1,1], [1,0,1,1,1], [1,1,0,1,1], [1,1,1,0,1], [1,1,1,1,0], [1,1,1,1,1]]:
            desired_outputs.append(1)
                    
        else:
            desired_outputs.append(0)

    manual_weights = [0.1, 0.3, 0.4, 0.3, 0.3, 1.0]

    perceptron = Perceptron(num_inputs = 5)
    perceptron.set_weights(manual_weights)
    print("Testing with manually set weights...")
    for i in range(len(training_inputs)):
        inputs = training_inputs[i][:]
        desired_output = desired_outputs[i]
        print("Input:", inputs, "-> Predicted Output:", perceptron.predict(inputs), ", Desired Output:", desired_output)

    trained_perceptron = Perceptron(num_inputs = 5)
    trained_perceptron.train(training_inputs, desired_outputs, rounds = 1000)
    print("\nTesting with trained weights...")
    for i in range(len(training_inputs)):
        inputs = training_inputs[i][:]
        desired_output = desired_outputs[i]
        print("Input:", inputs, "-> Predicted Output:", trained_perceptron.predict(inputs), ", Desired Output:", desired_output)
    
    print("\nFinal weights:", trained_perceptron.weights, "\n")
